### PR TITLE
Process control

### PR DESCRIPTION
**Pull request by phil-levis**
### Pull Request Overview

This pull request adds a process management console over the UART. It exposes 3 commands: help, list, stop, and start. Implementing this required a few changes to kernel inspection interfaces. In particular, the Introspection interface (which I renamed to KernelInfo, since it is not an introspection interface) is not very usable. By exposing AppIds, it forces the console to scan the process list twice: once to get the AppId from a name, then to find the process that has that AppId. So I modified Kernel to have an process_each_capability, a public method that allows iterating across processes but requires a ProcessManagementCapability.

This is not 100% polished -- just wanted to get it in people's hands so they can give feedback. For example, might want to change the boot sequence, code decomposition.

### Testing Strategy

This pull request was tested by running it on imix, loading two processes, and testing that stopping a process actually stops it.

### TODO or Help Wanted

This pull request still needs people to try it.

### Documentation Updated

- [X] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [ ] Ran `make formatall`.

Not yet.

### CODE REVIEW

**Comment by bradjc** [comment]
Will this work on top of `virtual_uart` with `Console` included as well?

File: boards/imix/src/components/process_console.rs [hunk 1]
```diff
@@ -0,0 +1,91 @@
+//! Component for ProcessConsole on the imix board.
+//!
+//! This provides one Component, ProcessConsoleComponent, which
+//! implements a command console for controlling processes over USART3
+//! (the DEBUG USB connector). It also attaches kernel debug output to this
+//! console (for panic!, print!, debug!, etc.).
+//!
+//! Usage
```
Comment (ppannuto) [inline boards/imix/src/components/process_console.rs]: I think the usage bit got copy/pasted from elsewhere

```suggestion
```

File: boards/imix/src/main.rs [hunk 2]
```diff
@@ -124,8 +124,13 @@ static mut PROCESSES: [Option<&'static kernel::procs::ProcessType>; NUM_PROCS] =
 #[link_section = ".stack_buffer"]
 pub static mut STACK_MEMORY: [u8; 0x2000] = [0; 0x2000];
 
+#[allow(dead_code)]
```
Comment (ppannuto) [inline boards/imix/src/main.rs]: It's probably emblematic of something a bit funkier in the overall pattern we have here, but I think this is better served by removing this and changing line 430 to be `imix.console.start()`. The `dead_code` allow is a bigger hammer that may hide other things in the future

File: capsules/src/process_console.rs [hunk 3]
```diff
@@ -0,0 +1,284 @@
+//! Provides implements a text console over the UART that allows
```
Comment (ppannuto) [inline capsules/src/process_console.rs]:
```suggestion
//! A text console over the UART that allows
```

File: capsules/src/process_console.rs [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+use core::str;
+use kernel::capabilities::ProcessManagementCapability;
+use kernel::common::cells::TakeCell;
+use kernel::hil::uart::{self, Client, UART};
+use kernel::introspection::KernelInfo;
+use kernel::Kernel;
+use kernel::ReturnCode;
+
+/// Syscall driver number.
+pub const DRIVER_NUM: usize = 0x00000001;
```
Comment (ppannuto) [inline capsules/src/process_console.rs]: Should this be the same driver number as a "regular" console, or a new one?

Comment (bradjc) [reply capsules/src/process_console.rs]: It's not currently a driver, so this can just be removed.

**Comment by phil-levis** [comment]
I've re-incorporated console, so now there is both console and process_console on imix.

I will fix up the comments and other feedback tonight.

**Comment by phil-levis** [comment]
Quanta/time slices are interchangable terms in scheduling. Feel free to change in quanta if you want.

File: capsules/src/process_console.rs [hunk 5]
```diff
@@ -221,7 +221,7 @@ impl<U: UART, C: ProcessManagementCapability> ProcessConsole<'a, U, C> {
                                 );
                             });
                         } else if clean_str.starts_with("list") {
-                            debug!(" PID    Name                Slices  Syscalls  Dropped Callbacks    State");
+                            debug!(" PID    Name               Quantas  Syscalls  Dropped Callbacks    State");
```
Comment (phil-levis) [inline capsules/src/process_console.rs]: The plural of quantum is quanta. Quantas is an airline. :)

Comment (ppannuto) [reply capsules/src/process_console.rs]: Sigh..

New Tock sponsor?! I'd leave it if they chipped in for some paid engineers :)

**Pull request merged**